File: general-parser/examples/parse.rs
use general_parser::{
	Adjacency, BinaryOperator, Configuration, Expression, ExpressionRepresentation,
	TernaryOperator, UnaryOperator,
};

fn main() {
	let arg = std::env::args().nth(1);

	if let Some("--interactive") = arg.as_deref() {
		run_interactive();
		return;
	}

	if let Some(path) = std::env::args().nth(1) {
		let source = std::fs::read_to_string(&path).unwrap();
		let (configuration, source) = extract_configuration_and_source(&source);

		eprintln!("{configuration:?}");

		let allocator = bumpalo::Bump::new();
		let expression: Expression<&str> =
			Expression::from_string(&source, &configuration, &allocator);

		let expression = ExpressionRepresentation(&expression);
		println!("{source}\n -> {expression}");
	} else {
		let configuration = Configuration {
			binary_operators: vec![
				BinaryOperator { representation: "*", precedence: 4 },
				BinaryOperator { representation: "+", precedence: 3 },
			],
			..Default::default()
		};

		let sources: &[&str] = &["(x (a * b) (d * 2 + e))", "(x (a b) (c d e))"];

		for source in sources {
			let allocator = bumpalo::Bump::new();
			let expression: Expression<&str> =
				Expression::from_string(source, &configuration, &allocator);
			let expression = ExpressionRepresentation(&expression);
			eprintln!("{source}\n -> {expression}");
		}
	}
}

fn run_interactive() {
	use std::io::{BufRead, stdin};
	let stdin = stdin();
	let mut buf = Vec::new();

	println!("start");

	for line in stdin.lock().lines().map_while(Result::ok) {
		if line == "close" {
			if !buf.is_empty() {
				eprintln!("no end to message {buf:?}");
			}
			break;
		}

		if line == "end" {
			let output = String::from_utf8_lossy(&buf);

			let (configuration, source) = extract_configuration_and_source(&output);

			// eprintln!("{configuration:?} {source}");

			{
				let allocator = bumpalo::Bump::new();
				let expression: Expression<&str> =
					Expression::from_string(&source, &configuration, &allocator);
				let expression = ExpressionRepresentation(&expression);
				println!("{expression}");
			}

			// if let Err(error) = out {
			//     println!("Error: {error:?}");
			// }

			println!("end");
			buf.clear();
			continue;
		}

		buf.extend_from_slice(line.as_bytes());
		buf.push(b'\n');
	}
}

fn extract_configuration_and_source(input: &str) -> (Configuration<'_>, &str) {
	if let Some((config, source)) = input.split_once("\n---") {
		let mut configuration = Configuration::default();
		for line in config.lines().map(str::trim_end) {
			let (adjacent, rest) = if let Some(rest) = line.strip_suffix(" (adjacent)") {
				(true, rest)
			} else {
				(false, line)
			};

			let (name, rest) = if let Some((name, rest)) = rest.split_once(':')
				&& name.trim_end().chars().all(is_identifier)
			{
				(name.trim_end(), rest)
			} else {
				("", rest)
			};

			let (syntax, precedence) = rest.split_once('#').unwrap_or((rest, "1"));
			let precedence: u8 = precedence.parse().expect("invalid precedence");
			let parts: Vec<_> = syntax.trim().split('_').map(str::trim).collect();

			match parts.as_slice() {
				["", first, part1, part2] => {
					configuration.postfix_ternary_operators.push(TernaryOperator {
						name,
						parts: (first, part1, part2),
						precedence,
					});
				}
				[first, part1, part2, ""] => {
					configuration.prefix_ternary_operators.push(TernaryOperator {
						name,
						parts: (first, part1, part2),
						precedence,
					});
				}
				["", binary_operator, ""] => {
					let operator = BinaryOperator { representation: binary_operator, precedence };
					if adjacent {
						configuration.adjacency =
							Some(Adjacency { operator, functions: Vec::default() });
					} else {
						configuration.binary_operators.push(operator);
					}
				}
				["", after] => {
					configuration
						.postfix_unary_operators
						.push(UnaryOperator { representation: after, precedence });
				}
				[before, ""] => {
					configuration
						.prefix_unary_operators
						.push(UnaryOperator { representation: before, precedence });
				}
				sequence => panic!("unknown sequence {sequence:?}"),
			}
		}
		(configuration, source.trim())
	} else {
		(Configuration::default(), input.trim())
	}
}

fn is_identifier(chr: char) -> bool {
	chr.is_alphanumeric() || matches!(chr, '_' | '$')
}

File: general-parser/javascript/parse.ts
import { BinaryOperator, Configuration, defaultConfiguration, parseExpression, parseExpressionPartial, printExpression } from "./index";
import { readFile } from "node:fs/promises";

if (process.argv.includes("--interactive")) {
	console.log("start");
	let buffer = "";
	for await (const line of console) {
		if (line == "close") break;

		if (line == "end") {
			const [partial, configuration, input] = extractConfigurationAndSource(buffer);
			if (partial) {
				const [expression, parsed] = parseExpressionPartial(input, configuration, partial.before);
				// TODO conversion here
				console.log(`parsed ${parsed} bytes`);
				console.log(printExpression(expression));
			} else {
				const expression = parseExpression(input, configuration);
				console.log(printExpression(expression));
			}
			console.log("end");
			buffer = "";
			continue
		}

		buffer += line;
		buffer += "\n";
	}

	// console.log("finished");
} else {
	const firstArgument = process.argv[2];
	if (firstArgument) {
		const buffer = await readFile(firstArgument);
		const [partial, configuration, input] = extractConfigurationAndSource(buffer.toString());
		console.log(configuration);
		if (partial) {
			const [expression, parsed] = parseExpressionPartial(input, configuration, partial.before);
			// TODO conversion here
			console.log(`parsed ${parsed} bytes`);
			console.log(printExpression(expression));
		} else {
			const expression = parseExpression(input, configuration);
			console.log(printExpression(expression));
		}
	} else {
		const buffer = "(x (a b) (c d))";
		const [_partial, configuration, input] = extractConfigurationAndSource(buffer);
		const expression = parseExpression(input, configuration);
		console.log(printExpression(expression));
	}
}

type Partial = { before: string | null } | null;

function extractConfigurationAndSource(input: string): [Partial, Configuration, string] {
	const configuration: Configuration = defaultConfiguration();

	if (input.includes("\n---")) {
		const [cfg, source] = input.split("\n---");
		let partial: Partial = null;

		for (let line of cfg.split("\n")) {
			let adjacency = false;

			if (line.startsWith("partial")) {
				const next = line.slice("partial".length).trim();
				const before = next.startsWith("upto ") ? next.slice("upto ".length) : null;
				partial = { before };
				continue;
			}

			if (line.endsWith(" (function)")) {
				const func = line.slice(0, - " (function)".length);
				if (!configuration.adjacency) throw new Error("Adjacency needed to register functions");
				configuration.adjacency.functions.push(func);
				continue;
			}

			if (line.endsWith(" (adjacent)")) {
				adjacency = true;
				line = line.slice(0, - " (adjacent)".length);
			}

			let name = "", precedence = 1;
			{
				const parts = line.split(":");
				if (parts.length > 1 && parts[0].trimEnd().match(/^([a-zA-Z0-9-_]+)$/m)) {
					name = parts[0].trimEnd()
					line = line.slice(parts[0].length + 1);
				}
			}

			{
				const parts = line.split("#");
				if (parts.length > 1) {
					// @ts-ignore
					const item: string = parts.at(-1);
					precedence = parseInt(item);
					line = line.slice(0, -(item.length + 1))
				}
			}

			const parts = line.split("_").map(part => part.trim());

			if (parts.length === 4) {
				if (parts[0] === "") {
					// @ts-ignore
					const items: [string, string, string] = parts.slice(1, 4);
					configuration.postfix_ternary_operators.push({
						name,
						parts: items,
						precedence
					});
				} else {
					// @ts-ignore
					const items: [string, string, string] = parts.slice(0, 3);
					configuration.prefix_ternary_operators.push({
						name,
						parts: items,
						precedence
					});
				}
			} else if (parts.length === 3) {
				const representation = parts[1];
				const operator: BinaryOperator = { representation, precedence };
				if (adjacency) configuration.adjacency = { operator, functions: [] };
				else configuration.binary_operators.push(operator);
			} else if (parts.length === 2) {
				if (parts[0] === "") {
					const representation = parts[1];
					configuration.postfix_unary_operators.push({
						precedence,
						representation
					});
				} else {
					const representation = parts[0];
					configuration.prefix_unary_operators.push({
						precedence,
						representation
					});
				}
			} else {
				throw new Error("Unknown operator");
			}
		}

		return [partial, configuration, source.trim()];
	} else {
		return [null, configuration, input.trim()];
	}
}
